### ISSUE TITLE
close(s::Session) waits forever 💤

### ISSUE THREAD

**Issue by samoconnor** [JuliaWeb/GnuTLS.jl#21]
close(s::Session) calls gnutls_bye(s.handle, GNUTLS_SHUT_RDWR).
This causes gnutls to "wait for the peer to reply with the same message" -- http://dev.man-online.org/man3/gnutls_bye

I'm using Requests.jl to talk to https://iam.amazonaws.com.
gnutls_bye hangs forever.

The patch below makes it work.
I don't really know enough about how GnuTLS.jl is intended to work to be sure this fix is a good one, or if AWS is "doing the wrong thing".
However, I think it is fair to say that being compatible with AWS is a useful thing.

My code calling Requests.jl is here: https://github.com/samoconnor/OCAWS.jl/blob/master/src/http.jl#L42

```patch
--- a/src/GnuTLS.jl
+++ b/src/GnuTLS.jl
@@ -88,12 +88,12 @@ const GNUTLS_SHUT_WR = 1


 free_session(s::Session) = ccall((:gnutls_deinit,gnutls),Void,(Ptr{Void},),s.handle)
-isopen(s::Session) = (isopen(s.read) || isopen(s.write))
+isopen(s::Session) = (s.open && (isopen(s.read) || isopen(s.write)))

 function close(s::Session)
        ret::Int32 = 0
        try # The remote might very well simply shut the stream rather than acknowledge the closure
-               ret = ccall((:gnutls_bye,gnutls), Int32, (Ptr{Void},Int32), s.handle, GNUTLS_SHUT_RDWR)
+               ret = ccall((:gnutls_bye,gnutls), Int32, (Ptr{Void},Int32), s.handle, GNUTLS_SHUT_WR)
        catch e
                if !isa(e,EOFError)
                        rethrow()
```

**Comment by samoconnor** [comment]
bump...

**Comment by sbromberger** [comment]
@samoconnor this may be related to https://github.com/JuliaWeb/Requests.jl/issues/42 - in any case, I think it's safe to say that GnuTLS is not in a stable state at the moment.

**Comment by sbromberger** [comment]
@samoconnor For what it's worth, your patch appears to fix https://github.com/JuliaWeb/Requests.jl/issues/42. This looks promising. Can you do a `Pkg.checkout("GnuTLS","sab/fixup1")` and `Pkg.checkout("Requests","sab/fixup")` and let me know whether you're getting better results?

**Comment by IainNZ** [comment]
@sbromberger I'm happy to report that I was able to run PkgEval with no glitches. Maybe make a new PR for GnuTLS with your changes (rebased to a single commit?) and we'll try and get Travis green. Tag that, get Requests green, then tag Requests.

**Comment by sbromberger** [comment]
Merged #36. Will tag shortly.

**Comment by sbromberger** [comment]
Tagged. Will close this out. I implemented this specific change (`GNUTLS_SHUT_RDWR` to GNUTLS_SHUT_WR`) in the PR.
